app/src/main/java/com/kks/nimblesurveyjetpackcompose/util/extensions/NetworkExtension.kt

```diff
@@ -0,0 +1,74 @@
+@file:Suppress("UnusedPrivateMember")
+package com.kks.nimblesurveyjetpackcompose.util.extensions
+
+import com.kks.nimblesurveyjetpackcompose.model.ResourceState
+import com.kks.nimblesurveyjetpackcompose.model.response.CustomErrorResponse
+import com.squareup.moshi.JsonDataException
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import kotlinx.coroutines.*
+import retrofit2.HttpException
+import retrofit2.Response
+import java.io.IOException
+import java.net.UnknownHostException
+
+private const val API_WAIT_TIME = 7000L
+const val UNKNOWN_ERROR = "Unknown error"
+const val SUCCESS_WITH_NULL_ERROR = "Success with null error"
+const val UNKNOWN_ERROR_MESSAGE = "Unknown error message"
+const val NETWORK_ERROR = "Network Error"
+
+const val EMAIL_EMPTY_MESSAGE = "Email cannot be empty"
+const val PASSWORD_EMPTY_MESSAGE = "Password cannot be empty"
+const val SUCCESS_MESSAGE = "success"
+
+private const val HTTP_ERROR_START = 400
+private const val HTTP_ERROR_END = 499
+const val INDEX_OF_ERROR_CODE = 0
+
+/**
+ * Reference: https://medium.com/@douglas.iacovelli/how-to-handle-errors-with-retrofit-and-coroutines-33e7492a912
+ */
+@Suppress("TooGenericExceptionCaught")
+suspend fun <T> safeApiCall(
+    dispatcher: CoroutineDispatcher,
+    apiCall: suspend () -> T
+): ResourceState<T> {
+    return withContext(dispatcher) {
+        try {
+            withTimeout(API_WAIT_TIME) {
+                ResourceState.Success(apiCall.invoke())
+            }
+        } catch (throwable: Exception) {
+            when (throwable) {
+                is IOException -> ResourceState.NetworkError
+                is UnknownHostException -> ResourceState.NetworkError
+                is TimeoutCancellationException -> ResourceState.NetworkError
```

Comment:
Could be grouped together 😄 
```suggestion
                is IOException,
                is UnknownHostException,
                is TimeoutCancellationException -> ResourceState.NetworkError
```